Remove Python 3.9 support

Aniketsy opened the pull request:
#2534 

Removed python 3.9 support.

Please let me know if my approach or fix needs any improvements . I’m open to feedback and happy to make changes based on suggestions.
Thankyou !

Fokko commented:
@Aniketsy Thanks for working on this! Could you update this one as well:

https://github.com/apache/iceberg-python/blob/40521c81828df6e99de4cdb9468bfab12a93c8b8/pyproject.toml#L70-L73

Aniketsy commented:
@Fokko I’ve updated the changes. I had one query, is it okay to work on an issue if no PR has been linked or mentioned for it yet?

Aniketsy commented on pyproject.toml:
```diff
@@ -68,8 +67,7 @@ google-auth = { version = ">=2.4.0", optional = true }
 duckdb = { version = ">=0.5.0,<2.0.0", optional = true }
 ray = [
-  { version = "==2.10.0", python = "<3.9", optional = true },
-  { version = ">=2.10.0,<=2.44.0", python = ">=3.9", optional = true },
+  { version = ">=2.10.0,<=2.44.0", python = ">=3.10", optional = true },
 ]
```
Thanks! sure

Aniketsy commented on pyproject.toml:
```diff
@@ -68,8 +67,7 @@ google-auth = { version = ">=2.4.0", optional = true }
 ray = [
-  { version = "==2.10.0", python = "<3.9", optional = true },
-  { version = ">=2.10.0,<=2.44.0", python = ">=3.9", optional = true },
+  { version = ">=2.10.0,<=2.44.0", optional = true }
```
Thanks for correcting me, you mean `{ }` . I will update this

Fokko commented:
@Aniketsy It looks like we're one `poetry lock` short? Can you also pull in latest main?

Aniketsy commented:
yes, i will update .

kevinjqliu commented on pyproject.toml:
```diff
@@ -50,7 +49,7 @@ include = [
 
 [tool.poetry.dependencies]
-python = "^3.9.2, !=3.9.7"
+python = ">=3.10,<3.13"
```
```suggestion
python = ">=3.10"
```
nit: lets not define an upper bound restriction here, this will allow others to experiment with 3.13

Aniketsy replied:
sure , thanks!

Aniketsy commented:
@Fokko  I’m not sure if I added the `poetry.lock` correctly. I installed the dependencies and added the file directly , please let me know if I did something wrong.

kevinjqliu reviewed (commented):
there are a few more mentions of `3.9` 
- .github/workflows/pypi-build-artifacts.yml
- .github/workflows/svn-build-artifacts.yml
- mkdocs/docs/contributing.md

Aniketsy commented:
@kevinjqliu I've applied the changes.

kevinjqliu commented on mkdocs/docs/contributing.md:
```diff
@@ -234,7 +234,7 @@ Deprecated in 0.1.0, will be removed in 0.2.0. The old_property is deprecated. P
 For the type annotation the types from the `Typing` package are used.
 
-PyIceberg offers support from Python 3.9 onwards, we can't use the [type hints from the standard collections](https://peps.python.org/pep-0585/).
+PyIceberg offers support from Python 3.10 onwards, we can't use the [type hints from the standard collections](https://peps.python.org/pep-0585/).
```
i think we can revert this change. 

https://peps.python.org/pep-0585/ is supported by 3.9

kevinjqliu replied:
ah looks like this was find/replaced when we removed 3.8

https://github.com/apache/iceberg-python/commit/857abd09d5cd547b357b6eb66dd327fecec4c605#diff-c0ad53549c03c8e66a0e3574698ce24dfa722b54ca7a2fdfcb8d62ff0343ab73R205-R206

kevinjqliu replied:
i think we can remove this paragraph entirely then

kevinjqliu commented:
looks like there a conflict, could you rebase this PR with main? 

```
git fetch origin main
git rebase origin/main -Xtheirs
```

You will likely have to regen the poetry.lock file again
```
poetry lock
```

kevinjqliu commented:
oops looks like poetry doesnt like that theres no upperlimit, we can use `<4.0` then

Fokko commented:
@Aniketsy it looks like we need to run another `poetry lock`

Fokko commented:
@Aniketsy Please make sure to pull in latest main first 👍
